### PR TITLE
Improve spec.template.spec.restartPolicy description

### LINKED ISSUE #111829: The value of .spec.template.spec.restartPolicy in kubectl explain Deployment/DaemonSet/StatefulSet template prompts an error
### What happened?

When using the command kubectl explain Deployment/DaemonSet/StatefulSet.spec.template.spec.restartPolicy, prompt
````
Restart policy for all containers within the pod. One of Always, OnFailure,Never. Default to Always.
````
But the value is Never, when kubectl applies the template file, it prompts
````
The Deployment "deployment" is invalid: spec.template.spec.restartPolicy: Unsupported value: "Never": supported values: "Always"

````

### What did you expect to happen?

When using the command kubectl explain Deployment/DaemonSet/StatefulSet.spec.template.spec.restartPolicy, prompt
````
Restart policy for all containers within the pod. Only Always. Default to Always.
````
But the value is

### How can we reproduce it (as minimally and precisely as possible)?

 kubectl explain Deployment.spec.template.spec.restartPolicy

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
Client Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.2.981+985c9202ccd250", GitCommit:"985c9202ccd250a5fe22c01faf0d8f83d804b9f3", GitTreeState:"clean", BuildDate:"2022-08-06T08:53:16Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.2.981+985c9202ccd250", GitCommit:"985c9202ccd250a5fe22c01faf0d8f83d804b9f3", GitTreeState:"clean", BuildDate:"2022-08-06T08:53:16Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
Improve spec.template.spec.restartPolicy description to avoid user misunderstanding of usage methods.

#### Which issue(s) this PR fixes:

Fixes #111829

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mengjiao-liu** [comment]
Wait for 1.25 Code Freeze to end.
/hold

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -3102,6 +3102,10 @@ type PodSpec struct {
 	// Restart policy for all containers within the pod.
 	// One of Always, OnFailure, Never.
 	// Default to Always.
+	// When the RestartPolicy field is specified in the Pod Template
+	// for Deployment/ReplicaSet/DaemonSet/StatefulSet,
+	// only a .spec.template.spec.restartPolicy equal to Always is allowed,
+	// which is the default if not specified.
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:3108]: documentation about use inside other workload objects probably isn't appropriate to put on the generic PodSpec type

Comment (mengjiao-liu) [reply staging/src/k8s.io/api/core/v1/types.go:3108]: So what are some good suggestions to avoid user misunderstanding?

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go:3108]: we could make a generic statement about some uses of pod spec restricting the value further, but this shouldn't enumerate those uses

Comment (mengjiao-liu) [reply staging/src/k8s.io/api/core/v1/types.go:3108]: > we could make a generic statement about some uses of pod spec restricting the value further.

Do you mean to restrict here or in other places?  
https://github.com/kubernetes/kubernetes/blob/619f1cf55208212893504338f18c69f3dcfcdc66/staging/src/k8s.io/api/core/v1/types.go#L3063

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go:3108]: On the RestartPolicy field inside PodSpec, we could add a line like "In some contexts, only a subset of those values may be permitted."

Then in the StatefulSetSpec, DaemonSetSpec, ReplicaSetSpec, ReplicationControllerSpec, JobSpec, CronJobSpec types, on the Template field, we could add a line to the godoc indicating the additional restrictions on the restart policy for the template.

**Comment by mengjiao-liu** [comment]
Due to a recent arm injury, this PR has been delayed and is expected to be updated in 10 days.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by mengjiao-liu** [comment]
Rebased.

File: pkg/apis/apps/types.go [hunk 2]
```diff
@@ -180,6 +180,9 @@ type StatefulSetSpec struct {
 	// of the StatefulSet. Each pod will be named with the format
 	// <statefulsetname>-<podindex>. For example, a pod in a StatefulSet named
 	// "web" with index number "3" would be named "web-3".
+	// When the RestartPolicy field is specified in the Pod Template for StatefulSetSpec,
+	// only a .spec.template.spec.restartPolicy equal to Always is allowed,
+	// which is the default if not specified.
```
Comment (liggitt) [inline pkg/apis/apps/types.go:185]: perhaps abbreviate to `The only allowed template.spec.restartPolicy value is "Always".`

Comment (mengjiao-liu) [reply pkg/apis/apps/types.go]: Updated. Thanks.

File: staging/src/k8s.io/api/apps/v1beta1/types.go [hunk 3]
```diff
@@ -420,6 +420,10 @@ type DeploymentSpec struct {
 	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
 
 	// Template describes the pods that will be created.
+	// When the RestartPolicy field is specified in the Pod Template for Deployment,
+	// only a .spec.template.spec.restartPolicy equal to Always is allowed,
+	// which is the default if not specified.
+	// The only allowed template.spec.restartPolicy value is "Always".
```
Comment (liggitt) [inline staging/src/k8s.io/api/apps/v1beta1/types.go:426]: simplify to `The only allowed template.spec.restartPolicy value is "Always".`

File: staging/src/k8s.io/api/apps/v1beta2/types.go [hunk 4]
```diff
@@ -429,6 +430,9 @@ type DeploymentSpec struct {
 	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
 
 	// Template describes the pods that will be created.
+	// When the RestartPolicy field is specified in the Pod Template for Deployment,
+	// only a .spec.template.spec.restartPolicy equal to Always is allowed,
+	// which is the default if not specified.
```
Comment (liggitt) [inline staging/src/k8s.io/api/apps/v1beta2/types.go:435]: simplify to `The only allowed template.spec.restartPolicy value is "Always".`

File: staging/src/k8s.io/api/apps/v1beta2/types.go [hunk 5]
```diff
@@ -690,6 +694,9 @@ type DaemonSetSpec struct {
 	// The DaemonSet will create exactly one copy of this pod on every node
 	// that matches the template's node selector (or on every node if no node
 	// selector is specified).
+	// When the RestartPolicy field is specified in the Pod Template for DaemonSetSpec,
+	// only a .spec.template.spec.restartPolicy equal to Always is allowed,
+	// which is the default if not specified.
```
Comment (liggitt) [inline staging/src/k8s.io/api/apps/v1beta2/types.go:699]: simplify to `The only allowed template.spec.restartPolicy value is "Always".`

File: staging/src/k8s.io/api/core/v1/types.go [hunk 6]
```diff
@@ -4217,6 +4217,8 @@ type ReplicationControllerSpec struct {
 
 	// Template is the object that describes the pod that will be created if
 	// insufficient replicas are detected. This takes precedence over a TemplateRef.
+
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:4220]: drop the blank line to keep the godoc comment together

File: pkg/apis/batch/types.go [hunk 7]
```diff
@@ -295,6 +295,7 @@ type JobSpec struct {
 	ManualSelector *bool
 
 	// Describes the pod that will be created when executing a job.
+	// The allowed template.spec.restartPolicy value is "Never" or "OnFailure".
```
Comment (liggitt) [inline pkg/apis/batch/types.go:298]:
```suggestion
	// The only allowed template.spec.restartPolicy values are "Never" or "OnFailure".
```

File: staging/src/k8s.io/api/batch/v1/types.go [hunk 8]
```diff
@@ -277,6 +277,7 @@ type JobSpec struct {
 	ManualSelector *bool `json:"manualSelector,omitempty" protobuf:"varint,5,opt,name=manualSelector"`
 
 	// Describes the pod that will be created when executing a job.
+	// The allowed template.spec.restartPolicy value is "Never" or "OnFailure".
```
Comment (liggitt) [inline staging/src/k8s.io/api/batch/v1/types.go:280]:
```suggestion
	// The only allowed template.spec.restartPolicy values are "Never" or "OnFailure".
```

File: staging/src/k8s.io/api/apps/v1beta1/types.go (hunk 3 above)
Comment (mengjiao-liu) [reply staging/src/k8s.io/api/apps/v1beta1/types.go]: 😅 I missed it. Updated. Thanks.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 6 above)
Comment (mengjiao-liu) [reply staging/src/k8s.io/api/core/v1/types.go]: Done.

**Pull request merged**